### PR TITLE
Fix multiple array issues in floating-point error estimation.

### LINKED ISSUE #332: Fix issues with arrays in error estimation
A number of grave issues regarding arrays exist currently. They are enumerated below:

1.  Duplicate and delta arrays are not initialized and allocated properly. 
2.  Array delta values are sometimes added twice.

### PR DESCRIPTION
**Pull request by grimmmyshini**
Possibly closes #332. 

**DO NOT** merge till #333 is fixed.  I intend to have some execution tests as well here.

### CODE REVIEW

**Comment by vgvassilev** [comment]
@grimmmyshini, can you rebase this PR?

File: lib/Differentiator/ErrorEstimator.cpp [hunk 1]
```diff
@@ -8,11 +8,33 @@
 using namespace clang;
 
 namespace clad {
+  QualType getUnderlyingArrayType(QualType baseType) {
+    if (baseType->isArrayType()) {
+      return baseType->getAsArrayTypeUnsafe()->getElementType();
```
Comment (vgvassilev) [inline lib/Differentiator/ErrorEstimator.cpp:13]: Maybe we can use `ASTContext::getBaseElementType`?

**Review by vgvassilev — COMMENTED** [review]
Can you rebase, rerun clang-format and squash?

File: lib/Differentiator/ErrorEstimator.cpp (hunk 1 above)
Comment (grimmmyshini) [reply lib/Differentiator/ErrorEstimator.cpp:13]: An object of ASTContext is not available in this function.

Comment (vgvassilev) [reply lib/Differentiator/ErrorEstimator.cpp:13]: Can't you pass it as an argument?

Comment (grimmmyshini) [reply lib/Differentiator/ErrorEstimator.cpp]: Okay, I can try that.

Comment (grimmmyshini) [reply lib/Differentiator/ErrorEstimator.cpp]: @vgvassilev done!

**Review by vgvassilev — APPROVED** [review]

**Review by vgvassilev — APPROVED** [review]
Apologies for the delay. LGTM!

**Pull request merged**